File: Find whether path exist - GFG/find-whether-path-exist.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends


class Solution
{
    public:
   //Function to find whether a path exists from the source to destination.
   bool is_Possible(vector<vector<int>>& grid) 
   {
        int dx[4]={1,-1,0,0};
   int dy[4]={0,0,1,-1};
       queue<pair<int,int>>q;
       int m=grid.size();
       int n=grid[0].size();
       for(int i=0;i<m;i++){
           for(int j=0;j<n;j++){
               if(grid[i][j]==1){
                 q.push({i,j});
                 break;
               }
           }
       }
       while(!q.empty()){
           auto p=q.front();
           q.pop();
           int a=p.first;
           int b=p.second;
           if(grid[a][b]==2)
              return 1;
              grid[a][b]=0;
             for(int i=0;i<4;i++){
                 int x=a+dx[i];
                 int y=b+dy[i];
             //   cout<<x<<y;
   
                
                 if(x>=0 && y>=0 && x<m && y<n && (grid[x][y]==3||grid[x][y]==2)){
                     if(grid[x][y]==2)
                         return 1;
                   else if(grid[x][y]==3)
                        q.push({x,y});
                 }
                 
             }
       }
   
       return 0;
       //code here
   }
};

// { Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<vector<int>>grid(n, vector<int>(n, -1));
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				cin >> grid[i][j];
			}
		}
		Solution obj;
		bool ans = obj.is_Possible(grid);
		cout << ((ans) ? "1\n" : "0\n");
	}
	return 0;
}  // } Driver Code Ends

File: Find whether path exist - GFG/find-whether-path-exist.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            int[][] grid = new int[n][n];
            for(int i = 0; i < n; i++){
                String[] S = br.readLine().trim().split(" ");
                for(int j = 0; j < n; j++){
                    grid[i][j] = Integer.parseInt(S[j]);
                }
            }
            Solution obj = new Solution();
            boolean ans = obj.is_Possible(grid);
            if(ans)
                System.out.println("1");
            else 
                System.out.println("0");
        }
    }
}
// } Driver Code Ends



class Solution
{
    //Function to find whether a path exists from the source to destination.
    public boolean dfs(int i, int j, boolean[][] visited, 
    int[][] grid){
        if(i < 0  || j < 0 || i >= grid.length || j >= grid[0].length){
            return false;
        }
        if(grid[i][j] == 2){
            return true;
        }
        
        if(visited[i][j] || grid[i][j] == 0)
        return false;
        
        visited[i][j] = true;
        boolean ans = false;
        ans |= dfs(i+1, j, visited, grid);
        ans |= dfs(i-1, j, visited, grid);
        ans |= dfs(i, j+1, visited, grid);
        ans |= dfs(i, j-1, visited, grid);
        return ans;
    }
    
    public boolean is_Possible(int[][] grid)
    {
        int V = grid.length;
        boolean[][] visited = new boolean[V+1][V+1];
        
        
        for(int i = 0; i<V; ++i){
            for(int j = 0; j<V; ++j){
                if(grid[i][j] == 1)
                return dfs(i,j,visited, grid);
            }
        }
        return false;
    }
}